server/src/main/java/org/elasticsearch/cluster/coordination/ElectionStrategy.java

```diff
@@ -105,10 +110,14 @@ public void beforeCommit(long term, long version, ActionListener<Void> listener)
         listener.onResponse(null);
     }
 
-    public boolean nodeMayWinElection(ClusterState lastAcceptedState, DiscoveryNode node) {
+    public NodeEligibility nodeMayWinElection(ClusterState lastAcceptedState, DiscoveryNode node) {
         final String nodeId = node.getId();
-        return lastAcceptedState.getLastCommittedConfiguration().getNodeIds().contains(nodeId)
+        if (lastAcceptedState.getLastCommittedConfiguration().getNodeIds().contains(nodeId)
             || lastAcceptedState.getLastAcceptedConfiguration().getNodeIds().contains(nodeId)
-            || lastAcceptedState.getVotingConfigExclusions().stream().noneMatch(vce -> vce.getNodeId().equals(nodeId));
+            || lastAcceptedState.getVotingConfigExclusions().stream().noneMatch(vce -> vce.getNodeId().equals(nodeId))) {
+            return new NodeEligibility(true, "");
```

Comment:
Since we now return the reason, it would be nice to split above condition and return different explanation for voting exclusion and getLastCommittedConfiguration.

Also should we take into account shutdown entries as well? I realize it is tricky if somebody create them for all nodes at once. Probably better not to, for now.

Reply:
I agree, but I think we should for this PR just add a reason for the shutdown case, and leave reason empty. That way we are not producing INFO log messages that are not really providing any concrete information. IMO, for INFO and above we should provide more detailed info, and ideally a test that makes sure the log message shows up when we want it to.

Reply:
I am imagining 
```
    public NodeEligibility nodeMayWinElection(ClusterState lastAcceptedState, DiscoveryNode node) {
        final String nodeId = node.getId();
        if (lastAcceptedState.getLastCommittedConfiguration().getNodeIds().contains(nodeId) == false
                && lastAcceptedState.getLastAcceptedConfiguration().getNodeIds().contains(nodeId) == false) {
            return new NodeEligibility(false, " not a voting node in the voting configuration");
        }
        if (lastAcceptedState.getVotingConfigExclusions().stream().anyMatch(vce -> vce.getNodeId().equals(nodeId))) {
            return new NodeEligibility(false, "node is scheduled for shutdown");
        }
        return new NodeEligibility(true, "");
    }
```
could provide more info as for why the node can not be elected. This should result in the same amount of log messages for when the node can not be elected. I agree that there is no reason to log anything if it can be elected.

Reply:
I think this would be the same as the existing logic, technically
```
    public NodeEligibility nodeMayWinElection(ClusterState lastAcceptedState, DiscoveryNode node) {
        final String nodeId = node.getId();
        if (lastAcceptedState.getLastCommittedConfiguration().getNodeIds().contains(nodeId) == false
            && lastAcceptedState.getLastAcceptedConfiguration().getNodeIds().contains(nodeId) == false) {
            if (lastAcceptedState.getVotingConfigExclusions().stream().anyMatch(vce -> vce.getNodeId().equals(nodeId))) {
                return new NodeEligibility(false, "node is scheduled for shutdown");
            }
        }
        return new NodeEligibility(true, "");
    }
```